geode-core/src/main/java/org/apache/geode/internal/cache/Oplog.java

```diff
@@ -5789,9 +5789,8 @@ boolean needsCompaction() {
       if (((rv / (double) rvHWMtmp) * 100) <= parent.getCompactionThreshold()) {
         return true;
       }
-    } else {
```

Comment:
This fix is incorrect. It only caused the empty oplog (created by the offline compaction) is not deleted by the recovery's auto compaction. 

I modified the dunit test a little bit and proved it. 

Unfortunately, there's no flag or status to show that a compaction is just finished. Even we want to use the empty oplog created by offline compaction, we need to careful arrange the logic. Such as:

change  needsCompaction to return int. 1: need to compact, -1: no need to compact, 0: there's one empty file,  only need to delete it.

But maybe there's better ideas.

Reply:
The new fix is good. 

Should use "} else if (hasNoLiveValues()) {" instead. 

And it's better to add a few comments here.

Reply:
Darrel and I discussed again. We might need to re-fix it the other way: 

Since the trouble is caused by offline compaction wrote totalCount=0 into krf, there's a potential bug that we restart with totalCount=0, but totalLiveCount=1, then with your current code change this oplog will never get a chance to compact until the only live entry is destroyed. 

So a better fix maybe: when restart and found totalCount=0, then update totalCount to be totalLiveCount. With this fix, you don't need to change needsCompaction() any more.

Reply:
I have changed `readKrf()` function. So that at the end of this function, when `totalCount` and `totalLiveCount` is updated after reading krf file, it will check if `totalCount` is `0` and `totalLiveCount` is greater than `0`, and update `totalCount` accordingly.